smdebug/profiler/profiler_config_parser.py

```diff
@@ -332,19 +339,83 @@ def write_tf_dataloader_flag(self, flag_filename):
+        if not self.profiling_enabled or not self.config.python_profiling_config.is_enabled():
+            return
+
+        return self.python_profiler.start_profiling(StepPhase.START)
+
+    def handle_step_start_python_profiling(self, mode: ModeKeys, current_step):
+        """Handle python profiling at the start of the step with the given mode and step by stopping python profiling
+        and starting python profiling again if python profiling stats should be saved for the current step.
+        """
+        return self._handle_step_python_profiling(StepPhase.STEP_START, mode, current_step)
+
+    def handle_step_end_python_profiling(self, mode: ModeKeys, current_step):
+        """Handle python profiling at the end of the step with the given mode and step by stopping python profiling
+        and starting python profiling again if python profiling stats should be saved for the current step.
+        """
+        return self._handle_step_python_profiling(StepPhase.STEP_END, mode, current_step)
+
+    def handle_forward_pass_end_python_profiling(self, mode: ModeKeys, current_step):
+        """Handle python profiling at the end of the forward pass with the given mode and step by stopping python
+        profiling and starting python profiling again if python profiling stats should be saved for the current step.
+        """
+        return self._handle_step_python_profiling(StepPhase.FORWARD_PASS_END, mode, current_step)
+
+    def start_post_hook_close_python_profiling(self):
+        """Start post-hook-close python profiling if python profiling is enabled.
+        """
+        if not self.profiling_enabled or not self.config.python_profiling_config.is_enabled():
+            return
+
+        return self.python_profiler.start_profiling(StepPhase.END)
+
+    def stop_post_hook_close_python_profiling(self):
+        """Stop post-hook-close python profiling if python profiling is enabled.
+        """
+        if not self.profiling_enabled or not self.config.python_profiling_config.is_enabled():
+            return
+
+        return self.python_profiler.stop_profiling(StepPhase.END)
+
 
 _profiler_config_parser = None
 
 
-def get_profiler_config_parser():
+def get_profiler_config_parser(framework: FRAMEWORK, create_new=False):
     """
-    Create a global profiler config parser object. This object loads the profiler config file and create
-    ProfilerConfig object.
+    Create a global profiler config parser object. This object loads the profiler config file and creates an
+    ProfilerConfig object if it doesn't already exist. Takes in the framework being used as an argument.
+
+    If `create_new` is set to `True`, then a new profiler config parser will be created regardless of whether one
+    already exists or not.
     :return:
     """
     global _profiler_config_parser
-    if _profiler_config_parser is None:
-        _profiler_config_parser = ProfilerConfigParser()
+    if create_new or _profiler_config_parser is None:
```

Comment:
nit: Rename `create_new` to `should_create_new_profiler_config_parser`

Reply:
What does `ProfilerConfigParser(framework=framework, create_new=create_new)` return for the following conditions?

```
| create_new | profiler_config |
|------------|-----------------|
| True       | True            |
| True       | False           |
| False      | True            |
| False      | False           |
```


I think there is scope to reduce this code to just

```
return ProfilerConfigParser(framework=framework, create_new=create_new)
```

Reply:
> I think there is scope to reduce this code to just
> 
> ```
> return ProfilerConfigParser(framework=framework, create_new=create_new)
> ```

no, the point is that if `create_new` is True and `profiler_config_parser` isn't initialized, we should create a new profiler config parser. otherwise we return the existing profiler config parser.

Reply:
If create_new is True we return a new ProfilerConfigParser regardless of the value of profiler_config_parser, right?

Reply:
But if `create_new` is False, then we shouldn't return a new ProfilerConfigParser object if an object already exists.